packages/grafana-ui/src/components/SelectNG/SelectNG.tsx

```diff
@@ -0,0 +1,104 @@
+import React from 'react';
+import Downshift from 'downshift';
+
+import { SelectNGProps } from './types';
+import {
+  itemToString,
+  renderDropdownButton,
+  DEFAULT_NO_OPTIONS_MESSAGE,
+  renderClearButton,
+  renderDefaultTrigger,
+  DEFAULT_PLACEHOLDER,
+} from './utils';
+import { SelectMenu, SelectMenuMessage } from './SelectMenu';
+import { useSelect } from './hooks';
+
+export function SelectNG<T>({
+  value,
+  onChange,
+  options,
+  placement = 'auto-start',
+  noOptionsMessage = DEFAULT_NO_OPTIONS_MESSAGE,
+  disabled,
+  clearable,
+  filterable,
+  placeholder = DEFAULT_PLACEHOLDER,
+}: SelectNGProps<T>) {
+  const { setPopperElement, triggerRef, filterOptions, popperProps, setPopperRef } = useSelect<T>({
+    options,
+    placement,
+  });
+
+  return (
+    <>
+      <Downshift
+        initialSelectedItem={value === null ? undefined : value}
+        itemToString={itemToString}
+        onSelect={item => {
+          onChange(item);
+        }}
+      >
+        {({
+          getInputProps,
+          getToggleButtonProps,
+          getMenuProps,
+          getItemProps,
+          isOpen,
+          highlightedIndex,
+          selectedItem,
+          inputValue,
+          openMenu,
+          closeMenu,
+          clearSelection,
+        }) => {
+          const dropdownButton = renderDropdownButton(triggerRef, isOpen, !!disabled, getToggleButtonProps);
+          const clearButton = clearable && selectedItem && renderClearButton(!!disabled, clearSelection);
+          const triggerElement = renderDefaultTrigger(
```

Comment:
As you can see Select and AsyncSelect have this code duplicated. Opinions on having something similar to SelectBase that would abstract away the Downshift?

Reply:
Personally I'd prefer breaking down this code into components instead of using render functions and abstracting this logic into respective components and then composing those components as needed.